src/shared/tools.h

```diff
@@ -219,9 +219,9 @@ inline char *newstring(const char *s)           { if(!s) s = ""; size_t l = strl
 #define loopv(v)     for(decltype((v).size()) i = 0; i<(v).size(); i++)
 #define loopvj(v)    for(decltype((v).size()) j = 0; j<(v).size(); j++)
 #define loopvk(v)    for(decltype((v).size()) k = 0; k<(v).size(); k++)
-#define loopvrev(v)  for(decltype((v).size()-1) i = (v).size()-1; i>=0; i--)
-#define loopvjrev(v) for(decltype((v).size()-1) j = (v).size()-1; j>=0; j--)
-#define loopvkrev(v) for(decltype((v).size()-1) k = (v).size()-1; k>=0; k--)
+#define loopvrev(v)  for(int i = (v).size() - 1; i >= 0; i--)
```

Comment:
Well, if anything, you should probably use `ssize_t` here.

Reply:
I had that idea too but I don't know how portable it is.

Reply:
It's a standard type. But the real fix would be to remove the macros and replace them with standard stuff. Since 2011 there is no more  acceptable reasons to keep those macros imo. I did so only to keep changes minimal, but they go in our way a lot too often.
We need to eradicate the `loop.*` macros. This is my opinion. 2 steps there: 1) do the job `cpp`  does and 2) use C++11 for loops when possible. 3) (yes I know) use iterators when possible.

The PR you're sending will certainly makes the current codebase more resilient (in practice and in what it looks like), but it will makes things harder for people that maintain it (more warnings about implicit conversions).

Reply:
I don't think removal of those macros is relevant in this PR. That's another issue.

> more warnings about implicit conversions

I didn't see any additional warnings when I compiled.

Reply:
I have not compiled it, but, I can see additional warnings, since we are trying to integrate STL containers there.
As you said, std::vector<>::size() returns a size_t. Thus, the change would trigger new casts,while unmasking some that should have happened.
In `short: int vector<>::size()` vs `size_t std::vector<T>::size() const`. You are true, there is indeed a problem when size() returns 0 for STL. I should have spotted that.
Now, the *real* fix, that would trigger no warning, would to replace that macro so that it would use reverse iterators. I'll admit, it's my fault, I should refactor everything correctly on 1st time. I should have 1st eradicated those stupid macros, with either proper std::for_each() instances, or with for( auto& el : element_list ) and then patch the thousands of lines of code, or...
Sorry.

Reply:
I think we just use `size_t` for the forward and `ssize_t` for the reverse versions for now. Ultimately, these macros should just be replaced anyway.